rust/src/lib/policy/capture.rs

```diff
@@ -0,0 +1,528 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::convert::TryFrom;
+
+use serde::{
+    ser::SerializeMap, Deserialize, Deserializer, Serialize, Serializer,
+};
+
+use crate::{ErrorKind, NetworkState, NmstateError};
+
+use super::{
+    iface::{get_iface_match, update_ifaces},
+    json::{get_value_from_json, value_retain_only, value_to_string},
+    route::{get_route_match, update_routes},
+    route_rule::{get_route_rule_match, update_route_rules},
+    token::{parse_str_to_tokens, NetworkCaptureToken},
+};
+
+pub(crate) const PROPERTY_SPLITTER: &str = ".";
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct NetworkCaptureRules {
+    pub rules: Vec<(String, NetworkCaptureCommand)>,
+}
+
+impl<'de> Deserialize<'de> for NetworkCaptureRules {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let map = serde_json::Map::<String, serde_json::Value>::deserialize(
+            deserializer,
+        )?;
+        let mut rules: Vec<(String, NetworkCaptureCommand)> = Vec::new();
+
+        for (k, v) in map.iter() {
+            rules.push((
+                k.to_string(),
+                NetworkCaptureCommand::deserialize(v)
+                    .map_err(serde::de::Error::custom)?,
+            ));
+        }
+        log::debug!("Parsed into rules {:?}", rules);
+        Ok(NetworkCaptureRules { rules })
+    }
+}
+
+impl Serialize for NetworkCaptureRules {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(self.rules.len()))?;
+        for (name, value) in &self.rules {
+            map.serialize_entry(&name, &value)?;
+        }
+        map.end()
+    }
+}
+
+impl NetworkCaptureRules {
+    pub(crate) fn execute(
+        &self,
+        current: &NetworkState,
+    ) -> Result<HashMap<String, NetworkState>, NmstateError> {
+        let mut ret = HashMap::new();
+        for (var_name, condition) in self.rules.as_slice() {
+            let matched_state = condition.execute(current, &ret)?;
+            log::debug!(
+                "Found match state for {}: {:?}",
+                var_name,
+                matched_state
+            );
+            ret.insert(var_name.to_string(), matched_state);
+        }
+        Ok(ret)
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.rules.is_empty()
+    }
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(deny_unknown_fields, try_from = "String", into = "String")]
+#[non_exhaustive]
+pub struct NetworkCaptureCommand {
```

Comment:
Let's have specific commands so they have specific fields and we don't mix the fact that a capture is piped there


```rust
pub trait Command {                                                             
    fn execute(&self) -> Result<NetworkState, NmstateError>;                    
}    

pub struct FilterCommand {                                                      
    input: Box<dyn Command>,                                                    
    path: Path,                                                                 
    value: Value,                                                               
}                                                                               
                                                                                
pub struct ReplaceCommand {                                                     
    input: Box<dyn Command>,                                                    
    path: Path,                                                                 
    value: Value,                                                               
}                                                                               
                                                                                
pub struct PathCommand {                                                        
    input: Box<dyn Command>,                                                    
    path: Path,                                                                 
}                                                                               
                                                                                
pub struct ExecuteCaptureCommand {                                              
    name: String,                                                               
    capture: Rv<Map<String, Box<dyn Command>>>,                                 
}                                                              
```

We can even use memoization and cache the Command::Execute result in the future.

Reply:
I don't know what this future looks like and why current code cause problem. Can you elaborate?

Reply:
Fair enough, I think we can refactor later on.